core/src/main/scala/ai/forestflow/utils/SourceStorageProtocols.scala

```diff
@@ -85,7 +93,44 @@ object SourceStorageProtocols extends StrictLogging {
   }
 
   case object S3 extends EnumVal {
-    def downloadDirectoryImpl(remotePath: String, localDirectory: File, fqrv: FQRV, sslVerify: Boolean): Unit = ???
+
+    private implicit val actorSystem: ActorSystem = ActorSystem("Application-s3-stream")
+    private implicit val executionContext: ExecutionContext = actorSystem.dispatcher
+    private implicit val materializer: ActorMaterializer = ActorMaterializer()
+
+    private val s3PathStylePathPattern = "/(.*)/(.*)".r
+
+    def downloadDirectoryImpl(remotePath: String, localDirectory: File, fqrv: FQRV, sslVerify: Boolean): Unit = {
+
+      val uri = URI.create(remotePath)
+
+      val (bucket, bucketKey) = uri.getPath match {
+        case s3PathStylePathPattern(bucket, bucketKey) => (bucket, bucketKey)
+        case _ => throw new IllegalArgumentException(s"Invalid url $remotePath, bucket and/or key name information couldn't be extracted from the url's path")
+      }
+
+      val localFilePath = Paths.get(localDirectory.getAbsolutePath, bucketKey)
+
+      val s3Settings = S3Ext(actorSystem).settings
+        .withEndpointUrl(uri.getScheme + "://" + uri.getAuthority)
+        .withPathStyleAccess(true) //this is stop gap solution until we get access to virtual hosting of buckets
+
+      val s3File: Source[Option[(Source[ByteString, NotUsed], ObjectMetadata)], NotUsed] = scaladsl.S3
+        .download(bucket, bucketKey)
+        .withAttributes(S3Attributes.settings(s3Settings))
+
+      //an unsuccessful download operation like bucket and/or object not found would lead to removal of the file from the given local directory
+      val result: Future[IOResult] = s3File
+        .flatMapConcat {
+          case Some((data: Source[ByteString, _], metadata)) => data
+          case _ => Source.empty
+        }.runWith(FileIO.toPath(localFilePath))
+
+      result onComplete {
+        case Success(value) => logger.debug(s"s3 download complete and successfully saved the file locally, written ${value.count / (1024.0 * 1024.0)} mb")
+        case Failure(error) => logger.error(s"s3 download unsuccessful with: ${error.printStackTrace()}")
+      }
```

Comment:
The download function needs to either complete successfully by guaranteed the model has been downloaded or fail. This looks like the downloading of the file could actually continue to execute in another thread while the downloadDirectoryImpl function actually concludes so the caller would be oblivious to the fact that download hasn't finished yet.